ref_cache/http_parser.c

```diff
@@ -0,0 +1,623 @@
+    parser->val = NULL;
+    parser->user_agent = NULL;
+    parser->referrer = NULL;
+
+    parser->buffer = malloc(BUF_SZ);
+    if (parser->buffer == NULL) {
+        return -1;
+    }
+
+    parser->upstream = upstream;
+
+    return 0;
+}
+
+void cleanup_http_parser(Http_Parser *parser) {
+    if (parser->uri) {
+        free(parser->uri);
+        parser->uri = NULL;
+    }
+    if (parser->key) {
+        free(parser->key);
+        parser->key = NULL;
+    }
+    if (parser->val) {
+        free(parser->val);
+        parser->val = NULL;
+    }
+    if (parser->user_agent) {
+        free(parser->user_agent);
+        parser->user_agent = NULL;
+    }
+    if (parser->referrer) {
+        free(parser->referrer);
+        parser->referrer = NULL;
+    }
+    if (parser->buffer) {
+        free(parser->buffer);
+        parser->buffer = NULL;
+    }
+}
+
+static char * parser_get_line(Http_Parser *parser, size_t *len) {
+    static char line[BUF_SZ + 1];
+    size_t lpos;
+
+    /* Search for end of line */
+    while (parser->pos != parser->in) {
+        if (parser->buffer[parser->pos] == '\x0a') break;
+        parser->pos = (parser->pos + 1) & BUF_MASK;
+    }
+    if (parser->pos == parser->in) {
+        if (parser->used == BUF_SZ) {
+            parser->state = ERR_TOO_LARGE;
+            return NULL;
+        }
+        return NULL;
+    }
+
+    lpos = 0;
+    if (parser->pos < parser->out) {
+        memcpy(&line[lpos], parser->buffer + parser->out, BUF_SZ - parser->out);
+        lpos += BUF_SZ - parser->out;
+        parser->out = 0;
+    }
+    memcpy(&line[lpos],
+           parser->buffer + parser->out, parser->pos - parser->out);
+    lpos += parser->pos - parser->out;
+    line[lpos] = '\0';
+    if (lpos > 0 && line[lpos - 1] == '\x0d') line[--lpos] = '\0';
+    *len = lpos;
+    parser->pos = (parser->pos + 1) & BUF_MASK; /* Get past \n */
+    parser->out = parser->pos; /* Eat the data */
+    parser->used = (parser->out > parser->in
+                    ? BUF_SZ - parser->out + parser->in
+                    : parser->in - parser->out);
+    return line;
+}
+
+static Read_result parser_read_input(Http_Parser *parser, int fd) {
+    ssize_t res = 1;
+    int short_read = 0;
+
+    /* Read into circular buffer */
+
+    assert(parser->used < BUF_SZ);
+
+    if (parser->in >= parser->out) {
+        size_t to_read = BUF_SZ - parser->in - (parser->in == 0);
+        res = read(fd, parser->buffer + parser->in, to_read);
+        if (res < 0) {
+            if (errno != EAGAIN || errno != EWOULDBLOCK || errno != EINTR) {
+                fprintf(stderr, "Error from fd #%d : %s\n", fd, strerror(errno));
+                return READ_ERROR;
+            }
+            short_read = 1;
+        } else {
+            short_read |= (size_t) res < to_read;
+            parser->in = (parser->in + res) & BUF_MASK;
+            parser->used += (unsigned int) res;
+        }
+    }
+    if (res > 0 && (parser->in + 1) < parser->out) {
```

Comment:
These two conditionals both fail when in+1 == out.

```
    if (parser->in >= parser->out) {
    ...
    if (res > 0 && (parser->in + 1) < parser->out) {
```

As a consequence, it can enter an infinite loop of read-more.

I triggered it by giving it a carefully crafted range request:

```
curl -v -s -r `perl -e 'print "x"x1009'` localhost:8000/f3814841f1939d3ca19072d9e89f3fd7
```

Reply:
This bit has been rewritten to use `readv()`, to make it simpler.  It shouldn't get stuck any more.